crates/store/re_chunk/src/iter.rs

```diff
@@ -543,6 +544,469 @@ impl Chunk {
+                slice_as_buffer_native(component_name, array, component_offsets)
+            }
+        }
+    };
+}
+
+impl_buffer_native_type!(u8);
+impl_buffer_native_type!(u16);
+impl_buffer_native_type!(u32);
+impl_buffer_native_type!(u64);
+impl_buffer_native_type!(i8);
+impl_buffer_native_type!(i16);
+impl_buffer_native_type!(i32);
+impl_buffer_native_type!(i64);
+impl_buffer_native_type!(f32);
+impl_buffer_native_type!(f64);
+impl_buffer_native_type!(i128);
+
+/// The actual implementation of `impl_array_list_native_type!`, so that we don't have to work in a macro.
+fn slice_as_array_list_native<
+    'a,
+    const N: usize,
+    T: arrow2::types::NativeType + arrow::datatypes::ArrowNativeType,
+>(
+    component_name: ComponentName,
+    array: &'a dyn Arrow2Array,
+    component_offsets: impl Iterator<Item = (usize, usize)> + 'a,
+) -> impl Iterator<Item = Vec<&'a [[T; N]]>> + 'a
+where
+    [T; N]: bytemuck::Pod,
+{
+    let Some(inner_list_array) = array.as_any().downcast_ref::<Arrow2ListArray<i32>>() else {
+        if cfg!(debug_assertions) {
+            panic!("downcast failed for {component_name}, data discarded");
+        } else {
+            re_log::error_once!("downcast failed for {component_name}, data discarded");
+        }
+        return Either::Left(std::iter::empty());
+    };
+
+    let inner_offsets = inner_list_array.offsets();
+    let inner_lengths = inner_list_array.offsets().lengths().collect_vec();
+
+    let Some(fixed_size_list_array) = inner_list_array
+        .values()
+        .as_any()
+        .downcast_ref::<Arrow2FixedSizeListArray>()
+    else {
+        if cfg!(debug_assertions) {
+            panic!("downcast failed for {component_name}, data discarded");
+        } else {
+            re_log::error_once!("downcast failed for {component_name}, data discarded");
+        }
+        return Either::Left(std::iter::empty());
+    };
+
+    let Some(values) = fixed_size_list_array
+        .values()
+        .as_any()
+        .downcast_ref::<Arrow2PrimitiveArray<T>>()
+    else {
+        if cfg!(debug_assertions) {
+            panic!("downcast failed for {component_name}, data discarded");
+        } else {
+            re_log::error_once!("downcast failed for {component_name}, data discarded");
+        }
+        return Either::Left(std::iter::empty());
+    };
+
+    let size = fixed_size_list_array.size();
+    let values = values.values();
+
+    // NOTE: No need for validity checks here, `iter_offsets` already takes care of that.
+    Either::Right(component_offsets.map(move |(idx, len)| {
+        let inner_offsets = &inner_offsets.as_slice()[idx..idx + len];
+        let inner_lengths = &inner_lengths.as_slice()[idx..idx + len];
+        izip!(inner_offsets, inner_lengths)
+            .map(|(&idx, &len)| {
+                let idx = idx as usize;
+                bytemuck::cast_slice(&values[idx * size..idx * size + len * size])
+            })
+            .collect_vec()
```

Comment:
everything except `collect_vec` is not dependent on `N` -> we likely can save a lot of template instantiations by putting this into a second level utility method that only depends on `T` but not on `N`

Reply:
I don't think I'm ready to add even more layers and complex return types to all of this (we need the intermediary values too...) for something I don't really have a way to measure right now :/ Also this is private, no the number of instantiations is fixed, at least.